cpp/benchmarks/streaming/ndsh/q04.cpp

```diff
@@ -0,0 +1,612 @@
+
+/**
+ * @brief Run a derived version of TPCH-query 4.
+ *
+ * The SQL form of the query is:
+ * @code{.sql}
+ *
+ * SELECT
+ *     o_orderpriority,
+ *     count(*) as order_count
+ * FROM
+ *     orders
+ * where
+ *     o_orderdate >= TIMESTAMP '1993-07-01'
+ *     and o_orderdate < TIMESTAMP '1993-07-01' + INTERVAL '3' MONTH
+ *     and EXISTS (
+ *         SELECT
+ *             *
+ *         FROM
+ *             lineitem
+ *         WHERE
+ *             l_orderkey = o_orderkey
+ *             and l_commitdate < l_receiptdate
+ *     )
+ * GROUP BY
+ *     o_orderpriority
+ * ORDER BY
+ *     o_orderpriority
+ * @endcode{}
+ *
+ * The "exists" clause is translated into a left-semi join in libcudf.
+ */
+int main(int argc, char** argv) {
+    cudaFree(nullptr);
+
+    rapidsmpf::ndsh::FinalizeMPI finalize{};
+    cudaFree(nullptr);
+    // work around https://github.com/rapidsai/cudf/issues/20849
+    cudf::initialize();
+    auto mr = rmm::mr::cuda_async_memory_resource{};
+    auto stats_wrapper = rapidsmpf::RmmResourceAdaptor(&mr);
+    auto arguments = rapidsmpf::ndsh::parse_arguments(argc, argv);
+    auto ctx = rapidsmpf::ndsh::create_context(arguments, &stats_wrapper);
+    std::string output_path = arguments.output_file;
+    std::vector<double> timings;
+
+
+    for (int i = 0; i < arguments.num_iterations; i++) {
+        rapidsmpf::OpID op_id{0};
+        std::vector<rapidsmpf::streaming::Node> nodes;
+        auto start = std::chrono::steady_clock::now();
+        {
+            RAPIDSMPF_NVTX_SCOPED_RANGE("Constructing Q4 pipeline");
+            // Convention for channel names: express the *output*.
+            /* Lineitem Table */
+            // [l_commitdate, l_receiptdate, l_orderkey]
+            auto lineitem = ctx->create_channel();
+            // [l_orderkey]
+            auto filtered_lineitem = ctx->create_channel();
+            // [l_orderkey]
+            auto filtered_lineitem_shuffled = ctx->create_channel();
+
+            /* Orders Table */
+            // [o_orderkey, o_orderpriority]
+            auto order = ctx->create_channel();
+
+            // [o_orderkey, o_orderpriority]
+            // Ideally this would *just* be o_orderpriority, pushing the projection
+            // into the join node / dropping the join key.
+            auto orders_x_lineitem = ctx->create_channel();
+
+            // [o_orderpriority]
+            auto projected_columns = ctx->create_channel();
+            // [o_orderpriority, order_count]
+            auto grouped_chunkwise = ctx->create_channel();
+
+            nodes.push_back(read_lineitem(
+                ctx, lineitem, 4, arguments.num_rows_per_chunk, arguments.input_directory
+            ));
+            nodes.push_back(
+                filter_lineitem(ctx, lineitem, filtered_lineitem)
+            );  // l_orderkey
+            nodes.push_back(read_orders(
+                ctx, order, 4, arguments.num_rows_per_chunk, arguments.input_directory
+            ));
```

Comment:
The filter of the orders table is very selective. So I would recommend attempting to use a bloom filter to pre-filter `lineitem` before shuffling.

q03 has an example